### ISSUE TITLE
FediAct icon misplacement

### ISSUE THREAD

**Issue by Lartsch** [Lartsch/FediAct#28]
The current solution (icon with fixed position in the bottom right if external instances) is more of a WIP and does not display correctly in some browsers, especially mobile.

In the future, this will be moved into the sidebar of Mastodon.

But I am not sure yet where to place it on V3 instances. Any suggestions?

**Comment by Freeplayg** [comment]
I'm not sure about placing it directly into an element, because it can become an issue in other fedi platforms in the future as some have different UI's the user can choose from

**Comment by Freeplayg** [comment]
I'm thinking of possibly a small toast modal that appears for a few seconds to show that it's running and then goes away? The account switching could possibly just go into the extension popup (and moving settings under a submenu?)

**Comment by Freeplayg** [comment]
I quickly made this in inspect element for the popup lol
![image](https://user-images.githubusercontent.com/40742947/208791211-a22efa19-ad0b-4421-8ec4-c193d3869b82.png)

**Comment by Lartsch** [comment]
That's right, but it's a general "issue" with FediAct. The approach it uses requires quite some element identifiers to work reliable (as of now, different variants of v3 and v4 Mastodon). So at this point, I think it is neglectable that there would need to be additional identifiers for where to add the icon.

But in general, I'd also like to keep it simple.

I'm thinking about a generalized JSON template that could be added for each fediware and its versions:

- Software identifiers
    - That'd be the endpoints /nodeinfo/2.1, /nodeinfo/2.0, /nodeinfo/2.0.json, /nodeinfo/2.1.json or /api/v1/instance in most cases
- Toot identifiers for all views, including
    - Buttons and rules, e.g.
        - does the button need to be enabled first?
        - do we need to inject any buttons and if so, where?
    - Handles
    - Polls
    - Data IDs and hrefs
    - Asynchronous behaviour rules
- Profile identifiers
    - Follow button
    - Handle
- Account/directory card identifiers
    - Follow button (+ type, e.g. full button / icon only / ...)
- Supported resolve methods (302 redirects, unauthorized API endpoint, etc)
- FediAct icon inject point and type of injection
- etc.

FediAct would then process sites based on these generalized rules. But this will be a big update that is weeks away.

**Comment by Lartsch** [comment]
> I'm thinking of possibly a small toast modal that appears for a few seconds to show that it's running and then goes away? The account switching could possibly just go into the extension popup (and moving settings under a submenu?)

That'd sure be elegant. It depends on the implementation of the account switching. As of now, I thought it'd be best if users can configure on an per-site basis which account should be used and in that case it would be easier to do so in the content script and update a dictionary in the local browser storage from there.

But I guess this could also be done from the popup. Haven thought that through yet, thanks for your suggestions

**Comment by Lartsch** [comment]
> 

That looks great! Hmmm... I like that approach. I will have to think about how the per-site account setting could look like if done from the popup.

**Comment by Freeplayg** [comment]
> That looks great! Hmmm... I like that approach. I will have to think about how the per-site account setting could look like if done from the popup.

Do you mean per-site like if you're on mastodon.coffee, do actions on this account, if on mastodon.online, do it on another account? If so I'm not really sure what the point of that would be, since servers can contain many different things

**Comment by Lartsch** [comment]
> > That looks great! Hmmm... I like that approach. I will have to think about how the per-site account setting could look like if done from the popup.
> 
> Do you mean per-site like if you're on mastodon.coffee, do actions on this account, if on mastodon.online, do it on another account? If so I'm not really sure what the point of that would be, since servers can contain many different things

Yes, correct. A few users let me know that they would like this possibility.

But in the end, being able to switch accounts at all should be sufficient for the start. And I guess only some power users would benefit from a per-site setting, so I agree that, considering the effort for implementation, this is probably not worth it.

Also, it's just two clicks (popup > choose instance) to change, so even those power users could probably be satisfied.

Good input! I think we can go with the global account switcher and your proposal for the UI.

**Comment by Lartsch** [comment]
So we need some changes:
- Storage:
    - Use a dictionary to store configured home instances and their tokens
    - Use existing fediact_homeinstance key to store the currently set home instance (this way, there should be no settings loss on update because of changed key names)
    - Remove existing key fediact_token (since they will be stored in the dictionary)
- Background script
    - Loop through configured instances to grab the tokens and update dictionary accordingly
- Popup:
    - Implement your proposal for the UI
    - This also requires adjustments to popup.js
- Background / popup script communication
    - The popup should indicate if the token was successfully grabbed a) right after adding and b) when opening the popup
    - Home instances for which no token could be grabbed should be disabled so they can't be set as active home instance 
- Content script
    - Remove fixed icon
    - Show a toast with timeout to indicate 
    - Stop using fediact_token key in favor of the new dictionary
    - Implement required changes so instantly update the token to use when the home instance was switched from the popup
